shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/RouteCardData.kt

```diff
@@ -407,40 +368,83 @@ data class RouteCardData(
                                 LineOrRoute.Line(line, routes = emptySet())
                             } else LineOrRoute.Route(route)
                         } else null
                     }
-                    .mapNotNull { entry ->
-                        when (val key = entry.key) {
+                    .mapNotNull { (partialLineOrRoute, patternRoutePairs) ->
+                        when (partialLineOrRoute) {
                             is LineOrRoute.Line ->
                                 LineOrRoute.Line(
-                                    key.line,
-                                    routes = entry.value.mapNotNull { it.first }.toSet()
-                                ) to entry.value.mapNotNull { it.second }
-                            is LineOrRoute.Route -> key to entry.value.mapNotNull { it.second }
+                                    partialLineOrRoute.line,
+                                    routes = patternRoutePairs.mapNotNull { it.first }.toSet()
+                                ) to patternRoutePairs.mapNotNull { it.second }
+                            is LineOrRoute.Route ->
+                                partialLineOrRoute to patternRoutePairs.mapNotNull { it.second }
                             null -> null
                         }
                     }
                     .toMap()
+
+            return patternsByRouteOrLine
+        }
+
+        // build the map of LineOrRoute => Stop => Patterns.
+        // A stop is only included for a LineOrRoute if it has any patterns that haven't been seen
+        // at an earlier stop for that LineOrRoute.
+        private fun patternsGroupedByLineOrRouteAndStop(
+            globalData: GlobalResponse,
+            parentToAllStops: Map<Stop, Set<String>>
+        ): Map<LineOrRoute, Map<Stop, List<RoutePattern>>> {
+            val routePatternsUsed = mutableSetOf<String>()
+
+            val patternsGrouped =
+                mutableMapOf<LineOrRoute, MutableMap<Stop, MutableList<RoutePattern>>>()
+
+            globalData.run {
+                parentToAllStops.forEach { (parentStop, allStopsForParent) ->
+                    val patternsByRouteOrLine =
+                        patternsByRouteOrLine(allStopsForParent, globalData)
+                            // filter out a route if we've already seen all of its patterns
+                            .filterNot {
+                                routePatternsUsed.containsAll(
+                                    it.value.map { pattern -> pattern.id }.toSet()
+                                )
+                            }
+                    routePatternsUsed.addAll(
+                        patternsByRouteOrLine.flatMap { it.value }.map { it.id }
+                    )
+                    for ((routeOrLine, routePatterns) in patternsByRouteOrLine) {
+                        val routeStops = patternsGrouped.getOrPut(routeOrLine) { mutableMapOf() }
```

Comment:
shoot I think this could theoretically still be an issue where the GL `RouteOrLine.Line` might not exactly match at different stops. Going to rework a little bit more